### ISSUE TITLE
[Telemetry] Research GoLang logging libraries

### ISSUE THREAD

**Issue by Olshansk** [pokt-network/pocket#181]
## Objective

Research various logging libraries and compose a list of those that satisfy the requirements we need.

## Origin Document

In order to unblock the work in https://github.com/pokt-network/pocket/issues/173 (i.e. first iteration of logging grafana dashboards), we first need to pick a logging library before we update the codebase and build the dashboards.

## Goals

- [ ] Identify list of requirements we need for logging (e.g. log by level, log by module, etc...)
- [ ] Propose the research, findings and recommendations to the protocol team before logs are standardized across the codebase

## Deliverables

- [ ]  1-2 pager containing:
      - [ ] A list of various GoLang logging libraries
      - [ ] Logging requirements
      - [ ] Tradeoffs (complexity, benefits, features) of each logging library in a table or chart
- [ ] A presentation/discussion with the protocol team, after which a library is selected before moving forward

## Non-goals / Non-deliverables

- Standardizing logs across the codebase (i.e. no code)
- Building any dashboards or tooling
---

**Creator**: @Olshansk 
**Co-Owners**: @okdas

**Comment by Olshansk** [comment]
@okdas Please provides details/updates (even if not done, e.g. a page to notion) as you're working don't this.

**Comment by okdas** [comment]
There are [so many](https://awesome-go.com/logging/) different logging libraries! 

### What existing blockchain clients use

go-etherium: copied [log15](https://github.com/inconshreveable/log15) into the codebase and made some modifications
Erigon: also fork of `log15` https://github.com/ledgerwatch/log/tree/master/v3
AVAX: [uber-go/zap](https://github.com/uber-go/zap)
Algorand: [logrus](https://github.com/Sirupsen/logrus)
CosmosSDK: [zerolog](https://github.com/rs/zerolog)

### Comparison

`rs/zerolog` and `uber/zap` are libraries that avoid allocations and reflection, as a result they are incredibly fast. There is a traid off though in usability when going this route - because in order to avoid allocations one must specify a datatype for each field in logging event structure. `uber/zap` has a `SugaredLogger` mode when it still supports `printf`-like syntax while still being pretty quick.

`logrus` seems to be the most popular solution, also a pioneer library that was most adopted (probably still is most popular?), with very human-friendly API. As it is imported in many existing software, maintainers avoid API-breaking changes that would be necessary to improve performace. Perf is still decent though! 

`log15` is an old library that is no longer maintained. Hence it is forked by software that has been around for a while (such as geth). Its maintainer helped with development of github.com/go-kit/log, which is a part of `gokit` microservice framework. 

### Final thoughts

I think we need to either use `rs/zerolog` or `uber-go/zap` as they are not only most performant, but also being recommended by `logrus` maintainers. Though we don't necessarily need a very performant library for day-to-day operations, I can see a scenario when we want to dump all network messages client recieves into kafka/clickhouse or something similar for further analysis, and it'd be nice to have a smaller hit on performace.

Please visit following pages to check out the API:
* https://github.com/uber-go/zap#quick-start
* https://github.com/rs/zerolog#getting-started

A few notes I would like to add:
1. Let the loglevel be configurable via argument and/or environment variable, with INFO by default;
2. Similarly, it might be benificial to let the user switch between logging output types - human-readable and JSON. I'd keep JSON by default as this is preferred for logging aggregation tools, and we have loki on local setup to run queries against logs. 
3. We should utilize logging structures when it makes sense to provide additional context or make it easier to write queries. For example, when log line is related to particular block, it makes sense to include `BlockID` or `BlockHash` as one of the arguments.


@Olshansk

**Comment by Olshansk** [comment]
Thanks for the awesome summary @okdas! Let's present and discuss this at our team meeting tomorrow so everyone is on the same page, but otherwise it looks like we have a path forward.

+1 to the use-cases you described in the notes section.

_Note: we should also capture this in a github discussion or in a markdown so this research isn't lost :)_

**Comment by Olshansk** [comment]
@okdas Any chance you move the research of this research to a github discussion similar to https://github.com/pokt-network/pocket/discussions/213?

**Comment by okdas** [comment]
@Olshansk sure thing! https://github.com/pokt-network/pocket/discussions/214

Note I created a separate category "Architecture" - with intent to keep the discussions related to client/software architecture there. Lmk if you don't think this is a good approach.
